### PR TITLE
Send notifications as cloud events

### PR DESCRIPTION
**Pull request by josejulio**
# Cloud events

We (notifications) intend to start receiving cloud events instead of our current format. The schemas can be found on [event-schemas](https://github.com/redHatInsights/event-schemas) repository and java lib on [event-schemas-java](https://github.com/redHatInsights/event-schemas-java) repository.

Since we know Policies application well enough, we want to start migrating this application to sending cloud events to notifications.

Current work is behind a feature flag, so that we can test it on pre-prod environments  before releasing it.

## What does this means for clients using third party integrations?

Clients will still receive the old payload - at least for now - we have plans to provide some migration time and serve both schemas side by side. How we are doing it is on discussion, but probably a switch/flag to tell the Integrations to use old or new schema.

## Sample kafka message to notifications

Notes: ~~The null fields (`host_url`, `hostname`, `rhel_version`) are values that the schema deem as optional, but i guess it would be doable to populate (and mark as required at least for policies).~~
The lib is now configured to omit the `null` values. We can still add them (and probably should, as we have the info)

```json
{
   "id":"00f867af-11b5-4fce-a2df-4da0aeb1d580",
   "source":"urn:redhat:policies:insights:policies",
   "subject":"urn:redhat:subject:rhel_system:23cacdef82dcc2b928a439e224c75d3f",
   "type":"com.redhat.console.insights.policies.policy-triggered",
   "dataschema":"https://console.redhat.com/api/schemas/apps/policies/v1/policy-triggered.json",
   "data":{
      "policies":[
         {
            "condition":"facts.arch = \"x86_64\"",
            "description":"abc",
            "id":"8903121b-96ff-4b09-9801-e3be93da29e1",
            "name":"abc",
            "url":"https://my-console.com/insights/policies/policy/8903121b-96ff-4b09-9801-e3be93da29e1"
         }
      ],
      "system":{
         "check_in":"2023-02-15T15:20:27.484Z",
         "display_name":"Sides",
         "tags":[
            {
               "key":"owner",
               "namespace":"insights-client",
               "value":"some-owner"
            },
            {
               "key":"contact",
               "namespace":"insights-client",
               "value":"some-email"
            },
            {
               "key":"location",
               "namespace":"insights-client",
               "value":"somewhere-0"
            },
            {
               "key":"location",
               "namespace":"insights-client",
               "value":"somewhere-1"
            },
            {
               "key":"location",
               "namespace":"insights-client",
               "value":"somewhere-2"
            }
         ],
         "inventory_id":"23cacdef82dcc2b928a439e224c75d3f"
      }
   },
   "time":"2023-02-15T15:20:28.839327995",
   "specversion":"1.0",
   "redhataccount":"123456",
   "redhatorgid":"123456"
}
```

## Depends

- https://github.com/RedHatInsights/notifications-backend/pull/1680

### CODE REVIEW

File: src/main/java/com/redhat/cloud/policies/engine/process/EventProcessor.java [hunk 1]
```diff
@@ -133,4 +109,89 @@ private static boolean hasNotificationAction(Policy policy) {
         }
         return false;
     }
+
+    private List<Policy> fireAndGetPoliciesWithActions(Event event, List<Policy> enabledPolicies) {
```
Comment (gwenneg) [inline src/main/java/com/redhat/cloud/policies/engine/process/EventProcessor.java:113]:
```suggestion
    private List<Policy> getFiredPoliciesWithNotificationAction(Event event, List<Policy> enabledPolicies) {
```

File: src/test/java/com/redhat/cloud/policies/engine/process/EventProcessorTest.java [hunk 2]
```diff
@@ -89,21 +98,74 @@ void testEnabledPoliciesFoundAndAllPoliciesFiredAndSent() {
...
+            assertEquals(1, tags.stream().map(RHELSystemTag::getNamespace).collect(Collectors.toSet()).size()); // Only 1 different namespace
+
+            assertTrue(List.of("Contact", "Location").containsAll(tags.stream().map(RHELSystemTag::getKey).collect(Collectors.toList())));
+            assertTrue(List.of("spam@redhat.com", "Neuchatel", "Charmey").containsAll(tags.stream().map(RHELSystemTag::getValue).collect(Collectors.toList())));
+            assertTrue(List.of(DEFAULT_NAMESPACE).containsAll(tags.stream().map(RHELSystemTag::getNamespace).collect(Collectors.toList())));
+
+            assertPolicyIncludedInCloudEventPolicyTriggered(policy1, cloudEvent);
+            assertPolicyIncludedInCloudEventPolicyTriggered(policy2, cloudEvent);
+
+            eventProcessor.process(event);
```
[the remark this answers is not in the document]
Comment (josejulio) [reply src/test/java/com/redhat/cloud/policies/engine/process/EventProcessorTest.java:163]: I guess not

```suggestion
```

File: src/main/java/com/redhat/cloud/policies/engine/process/NotificationSender.java [hunk 3]
```diff
@@ -43,6 +47,8 @@ public class NotificationSender {
     @Inject
     MeterRegistry meterRegistry;
 
+    ConsoleCloudEventParser parser = new ConsoleCloudEventParser();
```
[the remark this answers is not in the document]
Comment (josejulio) [reply src/main/java/com/redhat/cloud/policies/engine/process/NotificationSender.java:50]: No, unless we modify the `ObjectMapper` - I tried with the quarkus one at first. I had some issues when converting an object to json.

https://github.com/RedHatInsights/event-schemas-java/blob/main/src/main/java/com/redhat/cloud/event/parser/ConsoleCloudEventParser.java#L111-L115

**Review by gwenneg — APPROVED** [review]
LGTM

**Pull request merged**